sklearn/preprocessing/_discretization.py

```diff
@@ -233,9 +245,28 @@ def fit(self, X, y=None):
                 '`subsample` must be used with `strategy="quantile"`.'
             )
 
+        elif self.strategy != "quantile" and sample_weight is not None:
+            raise ValueError(
+                "`sample_weight` was provided but it can be only used with"
+                f"strategy='quantile'. Got strategy={self.strategy!r} instead."
+            )
+
+        valid_encode = ("onehot", "onehot-dense", "ordinal")
+        if self.encode not in valid_encode:
+            raise ValueError(
+                "Valid options for 'encode' are {}. Got encode={!r} instead.".format(
+                    valid_encode, self.encode
+                )
+            )
+
         n_features = X.shape[1]
         n_bins = self._validate_n_bins(n_features)
 
+        if sample_weight is not None:
+            sample_weight = _check_sample_weight(
+                sample_weight, X, dtype=X.dtype, copy=True
```

Comment:
I don't think that we need a copy. We can add a test later to make sure that `sample_weight` is not modified after the call to `fit`.

Reply:
```suggestion
                sample_weight, X, dtype=X.dtype
```